oeps/oep-0050-hooks-extension-framework.rst

```diff
@@ -0,0 +1,288 @@
+return nothing. Filters on the other hand are passed data and are expected to
+return something. This something will be used by the platform code. The most
+common case is to be passed data and return the same of somewhat changed, but
+similar data. Other common case would be to receive data and raise an exception
+in accordance with the definition of the hook. Most likely to completely halt
+the process that would happen after the hook.
+
+The documentation for the available hooks at any point should be kept in-code at
+https://github.com/edx/edx-platform/tree/master/docs/guides. This location
+should contain a document that covers the design principles and details that did
+not land in the OEP as well as the final decisions made in the OEP. Also will
+index the list of hooks and their current status. It should not cover the
+history and all the previous states of the framework. It should reflect the
+latest way of thinking and examples of how to use the framework.
+
+Given the public promise made by the framework, the signature of the functions
+(input/output) of every hook should be maintained in a stable way. The same
+applies for additions or removals from the list of public triggers, which should
+follow a path of deprecation first and preferably be given an optional hook
+where the same context information or code execution path can be accessed.
+
+Registering more than one function for a hook should be possible. This creates
+two different interfaces given that filters are expected to return something and
+the platform should act accordingly. Due to this different nature, the mechanism
+by which a developer connects to a hook differs from events to filters.
+
+Events will be called via Django signals. There will be a new library that
+defines a list of semantically meaningful signals based on higher-level events
+and platform concepts. This library will be imported at the platform code and
+will, when possible, replace the current signals the platform already emits.
+Developers that want to use a hook, should also import this library at the
+plugins where they are writing the functions that connect to the event.
+
+For filters, the possibility of registering more than one function makes some
+sort of reducing routine for the results necessary. Therefore the proposal is to
+make them work as a pipeline, so the results of the previously executed function
+should be made available to the next function. Doing so makes the calling order
+of its registered functions important and thus it must be respected. This is not
+supported by Django signals and for this case, the mechanism to register
+functions will be the Django settings module.
+
+When possible, hooks should be compatible with the Async Event Messaging
+described in OEP-41.
+In terms of error handling, the implementation should allow the configuration of
+silent failures letting operators know the source of the error.
```

Comment:
[+1] If I understand this statement correctly, the implementation will have the means to debug issues (logging statements, etc) to help operators understand the root cause of performance and functional issues caused by hooks.

Reply:
Yes, but cautiously. By this I mean that we definitely need to put some tooling in place in the implementation to allow debugging and help operators know what is going on. However we also don't want to make the library implementing this bloated by too many extra tools and scaffolding.